### PR TITLE
Security fixes to address #604

### LINKED ISSUE #604: The changes in #593 make nginx run as root by default without privilege separation on platform family debian
## :ghost: Brief Description

#593 was supposed to address the issues i reported in #591 but only made things worse. Nginx is now running as root by default on `platform_family?('debian')` without privilege separation (not dropping privileges after binding port). This is a serious security issue.

## :pancakes: Cookbook version

12.0.6 and up

## :woman_cook:  Chef-Infra Version

17.10.3

## :tophat: Platform details

Ubuntu 20.04

## Steps To Reproduce

1. converge cookbook 12.0.6 or newer
2. observe all processes running as root

```
root        2918  0.0  0.1  56256  5608 ?        Ss   Jun15   0:00 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
root     2612257  0.0  0.1  56956  5524 ?        S    10:43   0:00 nginx: worker process
root     2612258  0.0  0.1  56956  5520 ?        S    10:43   0:00 nginx: worker process
```

## :police_car: Expected behavior

expected situation:

I expect the process user to be `www-data` on platform family debian

```
root         642  0.0  0.0  56020   276 ?        Ss   May19   0:00 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
www-data     644  0.0  0.0  56812  3720 ?        S    May19   6:21 nginx: worker process
www-data     645  0.0  0.0  56680  3008 ?        S    May19   0:00 nginx: worker process
```

## :heavy_plus_sign: Additional context

My initial security issue report: #591 
The change that was applied #593 

Even though #593 fixed the initial issue of the webserver being able to alter/remove configuration files, it only did it for platform family Debian (apart from creating a large security issue). On all other platforms than Debian the webserver will run as user:group`nginx:nginx` while all configuration objects are owned by `nginx:nginx` mode 640. So the main issue at hand here is that the configuration files should have a different owner than the process user.

So i propose:

debian family:

All config files owned by `root:root` mode 640 with process user `www-data:www-data`

the others:

All config files owned by `root:root` mode 640 with process user `nginx:nginx`

### PR DESCRIPTION
**Pull request by hrak**
Fix nginx running as root on Debian family
Fix webserver able to overwrite/delete config files
Add Inspec tests

# Description

This PR partly reverts #593 that caused nginx to run as root on Debian platform family. It also makes sure that all configuration files are written as `root:root` so the webserver has no permission to alter or delete its own configuration.

## Issues Resolved

#604
#591 

## Check List

- [x] A summary of changes made is included in the CHANGELOG under `## Unreleased`
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Review by jeffbyrnes — CHANGES_REQUESTED** [review]
Seems good, but a few of the integration tests are failing. Any chance it’s something that can be fixed with this PR?

**Comment by hrak** [comment]
> Seems good, but a few of the integration tests are failing. Any chance it’s something that can be fixed with this PR?

I briefly looked into it but it seems unrelated to the changes. Something with RHEL + nginx repo + missing libpcre2 dependency.

**Comment by hrak** [comment]
Maybe the easiest solution would be to install pcre2 before the installation of nginx, so turn it into something like:

```ruby
if platform_family?('amazon', 'fedora', 'rhel') {
  package 'pcre2'
  package_install_opts = '--disablerepo=* --enablerepo=nginx' 
}
```

**Review by jeffbyrnes — APPROVED** [review]
Looks like you’re right, these are unrelated failures. Still, you squashed a few!

I think this can be merged in, and these failures dealt with separately.

**Comment by hrak** [comment]
> Looks like you’re right, these are unrelated failures. Still, you squashed a few!
> 
> I think this can be merged in, and these failures dealt with separately.

Yeah, i found the issue with fedora-latest already. It seems to require `openssl1.1` to be installed. I will file a PR later to fix that.

**Comment by hrak** [comment]
> I can't test this myself now but does using the CentOS 9 repo rather than 8 fix the Fedora install issue?

I just tested this, and this works without `openssl1.1` so thats probably the better solution. Will update PR.

**Review by bmhughes — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [12.1.0](https://github.com/sous-chefs/nginx/releases/tag/12.1.0)

**Comment by jeffbyrnes** [comment]
Firefly meme GIF of character about to say something, pausing, starting again, then resting their chin in their hand in rueful agreement

File: resources/config.rb [hunk 1]
```diff
@@ -49,12 +49,12 @@
           default: lazy { node['hostname'] }
 
 property :owner, String,
-          description: 'File/folder group',
-          default: lazy { nginx_user }
+          description: 'File/folder user',
+          default: 'root'
```
Comment (damacus) [inline resources/config.rb:53]: @hrak haven't we changed it to run *as* root here? Or am I misreading?

Comment (hrak) [reply resources/config.rb:53]: The latter. The description already gives it away. Property `owner` is used for the ownership of config files and folders. `process_user` and `process_group` is whats used as the user/group the process runs as after dropping privileges.

Comment (jeffbyrnes) [reply resources/config.rb:53]: Ohhh ok ok, yes. This is why this looked good to me; it breaks out _config file ownership_ from _what user runs NGINX_ and hardens security of the config files.

Hopefully I didn’t miss something in my review though 😟

**Pull request merged**